File: Demo/SJAttributesFactory/ViewController.m
//
//  ViewController.m
//  SJAttributesFactory
//
//  Created by 畅三江 on 2017/11/6.
//  Copyright © 2017年 畅三江. All rights reserved.
//

#import "ViewController.h"
#import "SJAttributesFactory.h"

static NSString *UITableViewCellID = @"UITableViewCell";

@interface ViewController ()<UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UILabel *testLabel;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UILabel *tipsLabel;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *heightConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *widthConstraint;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self _setupViews];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSLog(@"");
    });
    
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        NSLog(@"");
    });
}

- (void)_setupViews {
    
    // test btn
    self.testLabel.layer.cornerRadius = 8;
    self.testLabel.layer.borderColor = [UIColor blackColor].CGColor;
    self.testLabel.layer.borderWidth = 0.6;
    self.testLabel.clipsToBounds = YES;
    self.testLabel.backgroundColor = [UIColor whiteColor];
    self.testLabel.numberOfLines = 0;
    self.testLabel.text = @"请选择 demo ";

    // table view
    [self.tableView registerClass:NSClassFromString(UITableViewCellID) forCellReuseIdentifier:UITableViewCellID];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    

}

#pragma mark -
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSAttributedString *attr = nil;
    NSString *tips = nil;
    switch (indexPath.row) {
        case 0: {
            tips = @"常用方法";
            attr = [NSAttributedString sj_UIKitText:^(id<SJUIKitTextMakerProtocol>  _Nonnull make) {
                make.font([UIFont boldSystemFontOfSize:20]);
                make.lineSpacing(8);
                make.append(@"叶秋笑了笑，抬手取下了衔在嘴角的烟头。");
                make.underLine(^(id<SJUTDecoration>  _Nonnull make) {
                    make.style = NSUnderlineStyleSingle;
                    make.color = [UIColor orangeColor];
                });
                make.strikethrough(^(id<SJUTDecoration>  _Nonnull make) {
                    make.style = NSUnderlineStyleSingle;
                    make.color = [UIColor orangeColor];
                });
                
                make.regex(@"叶秋").update(^(id<SJUTAttributesProtocol>  _Nonnull make) {
                    make.font([UIFont boldSystemFontOfSize:40]).textColor([UIColor purpleColor]);
                });
                
                make.regex(@"笑了笑").replaceWithText(^(id<SJUIKitTextMakerProtocol>  _Nonnull make) {
                    make.append(@"xiao了笑");
                });
                
                make.regex(@"抬手").replaceWithString(@"Tai手");
            }];
        }
            break;
        case 1: {
            tips = @"正则匹配";
            attr = [NSAttributedString sj_UIKitText:^(id<SJUIKitTextMakerProtocol>  _Nonnull make) {
                make.font([UIFont boldSystemFontOfSize:20]);
                make.append(@"@迷你世界联机 :@江叔 用小淘气耍赖野人#迷你世界#");
                
                make.regex(@"[@][^\\s]+\\s").update(^(id<SJUTAttributesProtocol>  _Nonnull make) {
                    make.textColor([UIColor purpleColor]);
                });
                
                make.regex(@"[#][^#]+#").update(^(id<SJUTAttributesProtocol>  _Nonnull make) {
                    make.textColor([UIColor orangeColor]);
                });
            }];

        }
            break;
        case 2: {
            tips = @"上下图文";
            attr = [NSAttributedString sj_UIKitText:^(id<SJUIKitTextMakerProtocol>  _Nonnull make) {
                make.appendImage(^(id<SJUTImageAttachment>  _Nonnull make) {
                    make.image = [UIImage imageNamed:@"sample2"];
                    make.bounds = CGRectMake(0, 0, 50, 50);
                });
                make.append(@"\n上下图文");
                make.alignment(NSTextAlignmentCenter).lineSpacing(8);
            }];
        }
            break;
        case 3: {
            tips = @"左右图文";
            attr = [NSAttributedString sj_UIKitText:^(id<SJUIKitTextMakerProtocol>  _Nonnull make) {
                make.append(@"  左右图文  ");
                make.font([UIFont boldSystemFontOfSize:18]).textColor([UIColor redColor]);
                make.appendImage(^(id<SJUTImageAttachment>  _Nonnull make) {
                    make.image = [UIImage imageNamed:@"sample2"];
                    make.bounds = CGRectMake(0, -18, 50, 50);
                });
            }];
        }
            break;
//        case 8: {
//            tips = @"测试";
//
//            attr = sj_makeAttributesString(^(SJAttributeWorker * _Nonnull make) {
////                大写 P 表示 Unicode 字符集七个字符属性之一：标点字符
////                L：字母；
////                M：标记符号（一般不会单独出现）；
////                Z：分隔符（比如空格、换行等）；
////                S：符号（比如数学符号、货币符号等）；
////                N：数字（比如阿拉伯数字、罗马数字等）；
////                C：其他字符
//
//                make.append(@"&&&&###@@$$$$");
//
//                make.regexp(@"^[\\p{S}|\\p{P}|\\p{M}|\\p{Z}]+$", ^(SJAttributesRangeOperator * _Nonnull make) {
//                    make.textColor([UIColor redColor]);
//                });
//
//                [self updateConstraintsWithSize:make.sizeByWidth(self.view.bounds.size.width - 80)];
//            });
//        }
//            break;
    }
    
    
    if ( !attr ) return;
    _tipsLabel.text = tips;
    _testLabel.attributedText = attr;
    [self updateConstraints];
    
    NSLog(@"------------- end -------------");
}

- (void)updateConstraints {
    CGSize size = sj_textSize(_testLabel.attributedText, self.view.bounds.size.width - 80, CGFLOAT_MAX);
    _widthConstraint.constant = size.width;
    _heightConstraint.constant = size.height;
    
    [UIView animateWithDuration:0.3 animations:^{
        [self.view layoutIfNeeded];
    }];
}

static CGSize sj_textSize(NSAttributedString *attrStr, CGFloat width, CGFloat height) {
    if ( attrStr.length < 1 )
        return CGSizeZero;
    CGRect bounds = [attrStr boundingRectWithSize:CGSizeMake(width, height) options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingUsesFontLeading context:nil];
    bounds.size.width = ceil(bounds.size.width);
    bounds.size.height = ceil(bounds.size.height);
    return bounds.size;
}

#pragma mark -
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 20;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UITableViewCellID forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"demo %zd", indexPath.row];
    return cell;
}

@end

File: Demo/SwiftVersion/ViewController.swift
//
//  ViewController.swift
//  SwiftVersion
//
//  Created by BlueDancer on 2018/1/28.
//  Copyright © 2018年 畅三江. All rights reserved.
//

import UIKit

private let reuseIdentifier = "Cell"

class ViewController: UIViewController {

    @IBOutlet weak var label: UILabel!
    
    @IBOutlet weak var tipsLabel: UILabel!
    
    @IBOutlet weak var tableView: UITableView!
    
    @IBOutlet weak var heightConstraint: NSLayoutConstraint!
    
    @IBOutlet weak var widthConstraint: NSLayoutConstraint!
    
    var demosM: [SJDemoInfo] = [SJDemoInfo]()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        label.layer.borderColor = UIColor.black.cgColor
        label.layer.borderWidth = 0.6
        label.numberOfLines = 0
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: reuseIdentifier)
        
        self.createExamples()
        tableView.reloadData()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    func createExamples() -> Void {
        for i in 0...3 {
            let model = SJDemoInfo.init()
            switch (i) {
            case 0:
                model.name = "\(i) 常用方法"
                                
                let text = NSAttributedString.sj.makeText({ (make) in
                    make.font(.boldSystemFont(ofSize: 20)).textColor(.black).lineSpacing(8)
                    
                    make.append(":Image - ")
                    make.append({ (make) in
                        make.image = UIImage.init(named: "sample2")
                        make.bounds = CGRect.init(x: 0, y: 0, width: 30, height: 30)
                    })
                    make.append("\n")
                    make.append(":UnderLine").underLine({ (make) in
                        make.style = .single
                        make.color = .green
                    })
                    make.append("\n")
                    make.append(":Strikethrough").strikethrough({ (make) in
                        make.style = .single
                        make.color = .green
                    })
                    make.append("\n")
                    make.append(":BackgroundColor").backgroundColor(.green)
                    
                    make.append("\n")
                    make.append(":Kern").kern(6)
                    
                    make.append("\n")
                    let shadow = NSShadow.init()
                    shadow.shadowColor = UIColor.red
                    shadow.shadowOffset = .init(width: 0, height: 1)
                    shadow.shadowBlurRadius = 5
                    make.append(":Shadow").shadow(shadow)
                    
                    make.append("\n")
                    make.append(":Stroke").stroke({ (make) in
                        make.color = .red
                        make.width = 1
                    })
                    
                    make.append("\n")
                    make.append("oOo").font(.boldSystemFont(ofSize: 25)).alignment(.center)
                    
                    make.append("\n")
                    make.append("Regular Expression")
                    make.regex("Regular").update({ (make) in
                        make.font(.boldSystemFont(ofSize: 25)).textColor(.purple)
                    })
                    make.regex("ss").replace("SS")
                    make.regex("on").replace({ (make) in
                        make.append("ON😆").textColor(.red).backgroundColor(.green).font(.boldSystemFont(ofSize: 30))
                    })
                });
                
                model.size = text.sj_textSize(forPreferredMaxLayoutWidth: self.view.bounds.size.width - 80)
                
                model.task = { return text }
                break;
            case 1:
                model.name = "\(i) 正则匹配"
                let text = NSAttributedString.sj.makeText({ (make) in
                    make.append("@迷你世界联机 :@江叔 用小淘气耍赖野人#迷你世界#")
                    
                    make.regex("[@][^\\s]+\\s").update({ (make) in
                        make.font(.boldSystemFont(ofSize: 25)).textColor(.purple)
                    })
                    
                    make.regex("[#][^#]+#").update({ (make) in
                        make.font(.boldSystemFont(ofSize: 25)).textColor(.purple)
                    })
                })
                
                model.size = text.sj_textSize(forPreferredMaxLayoutWidth: self.view.bounds.size.width - 80)
                model.task = { return text }
                break;
            case 2:
                model.name = "\(i) 上下图文"
                let text = NSAttributedString.sj.makeText({ (make) in
                    make.alignment(NSTextAlignment.center).lineSpacing(8)
                    make.append({ (make) in
                        make.image = UIImage.init(named: "sample2")
                        make.bounds = CGRect.init(x: 0, y: 0, width: 30, height: 30)
                    })
                    
                    make.append("\n999")
                })
                
                model.size = text.sj_textSize(forPreferredMaxLayoutWidth: self.view.bounds.size.width - 80)
                model.task = { return text }
                break;
            default:
                model.name = "\(i) 图文对齐"
                let text = NSAttributedString.sj.makeText { (make) in
                    make.font(UIFont.systemFont(ofSize: 14))
                    make.append("Top->")
                    make.append({ (make) in
                        make.image = UIImage.init(named: "sample2")
                        make.alignment = .top
                    })
                    
                    make.append("Center->")
                    make.append({ (make) in
                        make.image = UIImage.init(named: "sample2")
                        make.alignment = .center
                    })
                    
                    make.append("Bottom->")
                    make.append({ (make) in
                        make.image = UIImage.init(named: "sample2")
                        make.alignment = .bottom
                    })
                }
                
                model.size = text.sj_textSize(forPreferredMaxLayoutWidth: self.view.bounds.size.width - 80)
                model.task = { return text }
                break;
            }
            demosM.append(model)
        }
        
    }
    
    func _updateHeightConstraint(_ size: CGSize) -> Void {
        widthConstraint.constant = size.width
        heightConstraint.constant = size.height
        UIView.animate(withDuration: 0.25) {
            self.view.layoutIfNeeded()
        }
    }
    
}

extension ViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if ( nil != demosM[indexPath.row].task ) {
            self.tipsLabel.text = demosM[indexPath.row].name
            label.attributedText = demosM[indexPath.row].task!()
            self._updateHeightConstraint(demosM[indexPath.row].size!)
        }
    }
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return demosM.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
        cell.textLabel?.text = demosM[indexPath.row].name
        return cell;
    }
}
